src/id_allocator.rs

```diff
@@ -0,0 +1,201 @@
+    InvalidInput,
+}
+
+pub type Result = result::Result<u32, Error>;
+
+impl fmt::Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        use self::Error::*;
+
+        match *self {
+            Overflow => write!(f, "There are no more available ids in the specified range."),
+            InvalidInput => write!(f, "Specified id is not in the range."),
+        }
+    }
+}
+
+/// # Example
+///
+/// ```rust
+/// # use vm_allocator::IdAllocator;
+/// 
+/// let mut legacy_irq_allocator = IdAllocator::new(5, 23);
+/// 
+/// for _ in 1..10 {
+///     let _id = legacy_irq_allocator.allocate_id().unwrap();
+/// }
+/// let irq = 8;
+/// let freed_id_result = legacy_irq_allocator.free_id(irq);
+/// if freed_id_result.is_err() {
+///     println!("Failed to free the specified id: {:?}", freed_id_result);
+/// }
+/// ```
+
+/// Id allocator representation.
+pub struct IdAllocator {
+    // Begining of the range of IDs that we want to manage.
+    range_base: u32,
+    // First available id that was never allocated. 
+    pub next_id: u32,
+    // End of the range of IDs that we want to manage.
+    range_end: u32,
+    // Set of all freed ids that can be reused at subsequent allocations.
+    pub freed_ids: BTreeSet<u32>,
+}
+
+impl IdAllocator {
+    /// Creates a new instance of IdAllocator that will be used to manage the
+    /// allocation and release of ids from the interval specified by
+    /// `range_base` and `range_max`
+    pub fn new(range_base: u32, range_max: u32) -> Self {
+        IdAllocator {
+            range_base: range_base,
+            next_id: range_base,
+            range_end: range_max,
+            freed_ids: BTreeSet::<u32>::new(),
+        }
+    }
+
+    /// Checks if a specified id belongs to the range that this instance of
+    /// `IdAllocator` is managing and if it was not freed before.
+    fn check_id(&self, id: u32) -> Result {
+        // Check for out of range.
+        if self.range_base > id || id > self.range_end || self.freed_ids.contains(&id) {
+            return Err(Error::InvalidInput);
+        }
+        Ok(id)
+    }
+
+    /// Allocate an ID from the managed ranged.
+    pub fn allocate_id(&mut self) -> Result {
+        // If the set containing all freed ids is not empty we extract the
+        // first entry from that set and return it.
+        if self.freed_ids.len() != 0 {
+            let ret_value = *self.freed_ids.iter().next().unwrap();
+            self.freed_ids.remove(&ret_value);
+            return Ok(ret_value);
+        }
+        // If no id was freed before we return the next available id.
+        let id = self.next_id;
+        if id > self.range_end {
+            return Err(Error::Overflow);
+        }
+        // Prepare the next available id and return `Error::Overflow` if the
+        // addition results in an overflow.
+        self.next_id = self.next_id.checked_add(1).ok_or(Error::Overflow)?;
```

Comment:
Above two Overflow Error may be confusing to the user since it's hard to know whether the next_id is overflow or exceed the range_end.
Perhaps add another explicit error type for exceeding the range_end?

Reply:
Sure, good suggestion

Reply:
This logic is a bit problematic because it does not allow you to reserve the last ID in the range if the range happens to end at u32::MAX. Can we instead keep the `last_used` and increment it when a new allocation request comes in? Then the error would be return only for the operation that is problematic (in this case u32::MAX + 1).